cmd/soroban-rpc/internal/methods/get_events.go

```diff
@@ -333,28 +365,59 @@ func (h eventsRPCHandler) getEvents(request GetEventsRequest) (GetEventsResponse
 			limit = request.Pagination.Limit
 		}
 	}
+	endLedger := request.StartLedger + LedgerScanLimit
+
+	if request.EndLedger != 0 {
+		endLedger = min(request.EndLedger, endLedger)
+	}
+
+	end := db.Cursor{Ledger: endLedger}
+	cursorRange := db.CursorRange{Start: start, End: end}
+
+	// Check if requested start ledger is within stored ledger range
+	if start.Ledger < ledgerRange.FirstLedger.Sequence || start.Ledger > ledgerRange.LastLedger.Sequence {
+		return GetEventsResponse{}, &jrpc2.Error{
+			Code: jrpc2.InvalidRequest,
+			Message: fmt.Sprintf(
+				"startLedger must be within the ledger range: %d - %d",
+				ledgerRange.FirstLedger.Sequence,
+				ledgerRange.LastLedger.Sequence,
+			),
+		}
+	}
 
 	type entry struct {
-		cursor               events.Cursor
+		cursor               db.Cursor
 		ledgerCloseTimestamp int64
 		event                xdr.DiagnosticEvent
 		txHash               *xdr.Hash
 	}
 	var found []entry
```

Comment:
Maybe it's worth giving this some large capacity to preallocate in case there are a lot of results that will lead to array expansion/copies?

Reply:
e.g. using `limit`